### PR TITLE
Fixes #15490 - adding view_host filter and improving error messages

### PR DESCRIPTION
**Pull request by lzap**
Fixing both missing permissions and incorrect error message.

### CODE REVIEW

File: app/services/foreman/access_permissions.rb [hunk 1]
```diff
@@ -317,7 +317,7 @@
                                       :externalNodes, :pxe_config, :storeconfig_klasses, :auto_complete_search, :bmc,
                                       :runtime, :resources, :templates, :overview, :nics],
                                     :dashboard => [:OutOfSync, :errors, :active],
-                                    :unattended => [:template, :provision],
+                                    :unattended => [:template, :provision] + TemplateKind.all.map(&:name).map(&:to_sym),
```
Comment (ares) [inline app/services/foreman/access_permissions.rb]: avoid double map, `TemplateKind.all.map { |kind| kind.name.to_sym }`

Comment (ares) [inline app/services/foreman/access_permissions.rb]: also I'd prefer using some hard list (maybe just put this into constant https://github.com/theforeman/foreman/blob/develop/db/seeds.d/07-config_templates.rb#L9), 
- one reason is that if plugin introduces new kind it should also map it's own permission, this would allow to review any custom template kind with view_host permission
- second is that this is required in initializer https://github.com/theforeman/foreman/blob/develop/config/initializers/foreman.rb#L1 and it's good to avoid using AR there

**Comment by ares** [comment]
In redmine issue you've mentioned `review_templates` permissions, did you change your mind?

**Comment by lzap** [comment]
Yeah, changed my mind because it looks like `unattended` with `provision` already works okay. Unattended does not have permission checks enabled, it's only reviews what need it. I don't see any security risk there. Let me change the AR to static list. AFAIK the only plugin that adds a template kind is bootdisk currently, but there is no way of previewing bootdisk templates therefore we don't need to add the permission symbol there?

**Comment by lzap** [comment]
Fixed, what you think?

**Comment by lzap** [comment]
Fixed failing tests. FYI these were expecting incorrect behavior. If you test this, the controller errors out with "Undefined method: layout" error because actually layout is not usable with this particular controller.

File: app/services/foreman/access_permissions.rb [hunk 2]
```diff
@@ -312,12 +312,13 @@
     pc_ajax_actions = [:parameters]
     subnets_ajax_actions = [:freeip]
     tasks_ajax_actions = [:show]
+    template_kinds =  [:PXELinux, :PXEGrub, :iPXE, :provision, :finish, :script, :user_data, :ZTP]
```
Comment (ares) [inline app/services/foreman/access_permissions.rb]: could you deduplicate this list? https://github.com/theforeman/foreman/blob/develop/db/seeds.d/07-config_templates.rb#L9

just put it into a constant somewhere in TemplateKind and use it in both places?

**Comment by ares** [comment]
Other than the duplication issue seems all right, I'll continue with testing when it's resolved.

**Comment by ohadlevy** [comment]
@lzap bump?

**Comment by lzap** [comment]
Done!

``` diff
diff --git a/app/models/template_kind.rb b/app/models/template_kind.rb
index 7c1e5a9..d647c1c 100644
--- a/app/models/template_kind.rb
+++ b/app/models/template_kind.rb
@@ -6,4 +6,8 @@ class TemplateKind < ActiveRecord::Base
   has_many :os_default_templates
   validates :name, :presence => true, :uniqueness => true
   scoped_search :on => :name
+
+  def self.predefined
+    [:PXELinux, :PXEGrub, :iPXE, :provision, :finish, :script, :user_data, :ZTP]
+  end
 end
diff --git a/app/services/foreman/access_permissions.rb b/app/services/foreman/access_permissions.rb
index a732de6..9b4488a 100644
--- a/app/services/foreman/access_permissions.rb
+++ b/app/services/foreman/access_permissions.rb
@@ -318,13 +318,12 @@ Foreman::AccessControl.map do |permission_set|
     pc_ajax_actions = [:parameters]
     subnets_ajax_actions = [:freeip]
     tasks_ajax_actions = [:show]
-    template_kinds =  [:PXELinux, :PXEGrub, :iPXE, :provision, :finish, :script, :user_data, :ZTP]

     map.permission :view_hosts,    {:hosts => [:index, :show, :errors, :active, :out_of_sync, :disabled, :pending, :vm,
                                       :externalNodes, :pxe_config, :storeconfig_klasses, :auto_complete_search, :bmc,
                                       :runtime, :resources, :templates, :overview, :nics],
                                     :dashboard => [:OutOfSync, :errors, :active],
-                                    :unattended => [:template, :provision] + template_kinds,
+                                    :unattended => [:template, :provision] + TemplateKind.predefined,
                                      :"api/v1/hosts" => [:index, :show, :status],
                                      :"api/v2/hosts" => [:index, :show, :status, :vm_compute_attributes],
                                      :"api/v2/interfaces" => [:index, :show],
```

**Comment by ares** [comment]
Could you please also use `TemplateKind.predefined` in https://github.com/theforeman/foreman/blob/develop/db/seeds.d/07-provisioning_templates.rb (that's why I wanted you to extract it)

**Comment by ares** [comment]
Why did you choose to use class method instead of constant? This will be harder to extend if we need it (only with alias_method_chain), if it was an array in constant, one could push to that array.

**Comment by lzap** [comment]
Pushed, the change had to be lost on the way...

**Comment by ares** [comment]
Tested, works fine. Ready to merge pending tests.

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -219,6 +219,7 @@ def require_admin
   end
 
   def render_403
+    return head(:forbidden) unless respond_to?(:layout)
```
Comment (domcleal) [inline app/controllers/application_controller.rb]: This seems to break regular 403 pages for me, if I log in as a user and try to access a route that I don't have permission for.  What's `layout` meant to be in this context?

Comment (lzap) [inline app/controllers/application_controller.rb]: We use this method in unattended controller where there is no layout set, which causes `Undefined method: layout` instead of 403. Will take a look, thanks.

**Comment by lzap** [comment]
Updated the check in the `render_403` error. This is not nice, if you have better idea how to return head in case the controller has no default layout set, let me know.

**Comment by lzap** [comment]
FYI we have three of these, the only relevant is unattended I think as the two are from API:

```
app/controllers/unattended_controller.rb
2:  layout false
app/controllers/api/v2/home_controller.rb
5:      layout false
app/controllers/api/v2/tasks_controller.rb
4:      layout false
```

File: test/functional/smart_proxy_auth_test.rb [hunk 4]
```diff
@@ -63,6 +63,5 @@ def test_failed_authentication_with_invalid_basic_auth_credentials_in_web_ui_con
     get :externalNodes, :id => 123
 
     assert_response :forbidden
-    assert_template "common/403"
```
Comment (domcleal) [inline test/functional/smart_proxy_auth_test.rb:170]: Is this change necessary now?

**Comment by domcleal** [comment]
> Updated the check in the render_403 error. This is not nice, if you have better idea how to return head in case the controller has no default layout set, let me know.

I can't suggest a better solution off the top of my head, but you should add some functional tests around render_403 in any case, for the regular and this particular code path.

**Comment by lzap** [comment]
I can see two possible solutions: this or explicitly check for the controller name there. The issue with unattended controller is that it was designed to return HTML format. It's actually text. Maybe a third solution is to make it a `txt` format. But there might be dragons! :-D

**Comment by lzap** [comment]
Rewritten using MIME text format, added tests.

File: test/functional/smart_proxy_auth_test.rb (hunk 4 above)
Comment (lzap) [inline test/functional/smart_proxy_auth_test.rb:170]: Yes, this is no longer valid.

Comment (ares) [inline test/functional/smart_proxy_auth_test.rb:170]: reverting it does not add any failure and seems to work, I suppose we still render the template here since you didn't specify text format

Comment (lzap) [inline test/functional/smart_proxy_auth_test.rb:170]: Hmmm it changed when I modified to the different approach. Removed.

**Comment by domcleal** [comment]
@lzap this isn't mergeable, please rebase

**Comment by lzap** [comment]
Done

**Comment by domcleal** [comment]
It seems possible to request a template for a host that isn't permitted by the user's filters.  I set up a filter on view_hosts where name = foo.example.com, was able to view that host's templates but I'm also able to view templates for hosts that I don't have the view permission on.

The unattended controller is probably going to have to use the user's authorised scope to retrieve it when in spoof mode?

The 403 response does seem to be working when the view_hosts permission is missing, but why show this without a layout?  I think in spoof mode having a layout could be an improvement, since it's part of the UI.

**Comment by lzap** [comment]
Summer is knocking the doors, time to resurrect this PR :-)

> The unattended controller is probably going to have to use the user's authorised scope to retrieve it when in spoof mode?

I will add this, and few tests.

> The 403 response does seem to be working when the view_hosts permission is missing, but why show this without a layout? I think in spoof mode having a layout could be an improvement, since it's part of the UI.

Unattended controller does not have any layout (note `layout false`) and this sometimes fail with `There was no default layout for UnattendedController` error. This patch tries to avoid that.

**Comment by lzap** [comment]
@ares would you mind looking at the test - I don't understand why it fails.

**Comment by ares** [comment]
> @ares would you mind looking at the test - I don't understand why it fails.

I have no idea and I can't reproduce it locally, let's try re [test]

**Comment by domcleal** [comment]
There will be a couple of failures in the MySQL matrix of the last test when it completes, which should be ignored.  A slave locked up while executing the tests - sorry.

**Comment by ares** [comment]
[test]

**Comment by ares** [comment]
tests are green here, @lzap I think there are no pending actions from your side so I removed the WoC

File: app/controllers/unattended_controller.rb [hunk 5]
```diff
@@ -79,7 +79,12 @@ def get_host_details
...
-      head(:not_found) and return
+      if @spoof
+        render(:text => _("Host not found or unauthorized to view this host"), :status => :not_found)
+      else
+        head(:not_found)
+      end
+      return
     end
     unless @host.operatingsystem
       logger.error "#{controller_name}: #{@host.name}'s operating system is missing!"
```
Comment (domcleal) [inline app/controllers/unattended_controller.rb]: The other errors in this method should be handled in the same way as above, and to be honest you could simply render the same :text string whether it's a spoof or not which would simplify the code.

File: test/functional/unattended_controller_test.rb [hunk 6]
```diff
@@ -358,6 +358,47 @@ class UnattendedControllerTest < ActionController::TestCase
   end
 
   test 'host template provision URL can be generated from routes' do
-    assert_routing '/unattended/provision', {:controller => 'unattended', :action => 'host_template', :kind => 'provision'}
+    assert_routing '/unattended/provision', {:controller => 'unattended', :action => 'host_template', :kind => 'provision', :format => 'text'}
+  end
+
+  test "should not render a template to anonymous user" do
+    User.current = nil
```
Comment (domcleal) [inline test/functional/unattended_controller_test.rb]: This should be the default, and set in the teardown.

File: test/functional/unattended_controller_test.rb [hunk 7]
```diff
@@ -358,6 +358,47 @@ class UnattendedControllerTest < ActionController::TestCase
...
+    get :host_template, {:kind => 'PXELinux', :spoof => @rh_host.ip, :format => 'text'}
+    assert_response :redirect
+  end
+
+  test "should not render a template to user w/o email" do
+    User.current = FactoryGirl.create(:user)
+    User.current.auth_source.stubs(:authenticated?).returns(false)
```
Comment (domcleal) [inline test/functional/unattended_controller_test.rb]: Unused?

File: test/functional/unattended_controller_test.rb [hunk 8]
```diff
@@ -358,6 +358,47 @@ class UnattendedControllerTest < ActionController::TestCase
...
+    session = SETTINGS[:login] ? {:user => User.current.id, :expires_at => 5.minutes.from_now} : {}
```
Comment (domcleal) [inline test/functional/unattended_controller_test.rb]: This looks like `set_session_user`, just add a line to the top of it so you can pass a User into it instead of relying on fixtures:

```
def set_session_user(user = :admin)
  user = user.is_a?(User) ? user : users(user)
  SETTINGS.....
end
```

You then don't have to mess with User.current and are using the existing helper.

File: test/functional/unattended_controller_test.rb [hunk 9]
```diff
@@ -358,6 +358,47 @@ class UnattendedControllerTest < ActionController::TestCase
...
+    assert_response :unprocessable_entity
+  end
+
+  test "should not render a template to unauthorized user" do
+    User.current = FactoryGirl.create(:user, :with_mail)
+    User.current.auth_source.stubs(:authenticated?).returns(false)
+    session = SETTINGS[:login] ? {:user => User.current.id, :expires_at => 5.minutes.from_now} : {}
+    get :host_template, {:kind => 'PXELinux', :spoof => @rh_host.ip, :format => 'text'}, session
+    assert_response :forbidden
+    assert_template nil
```
Comment (domcleal) [inline test/functional/unattended_controller_test.rb]: assert_response?  The controller doesn't use templates/partials etc.

Comment (domcleal) [inline test/functional/unattended_controller_test.rb]: > assert_response

Sorry!  assert the contents of `@response.body`.

File: test/functional/unattended_controller_test.rb [hunk 10]
```diff
@@ -358,6 +358,47 @@ class UnattendedControllerTest < ActionController::TestCase
...
+    assert @response.body.include?("linux")
+  end
+
+  test 'should not render a template to user with valid filter' do
+    User.current = FactoryGirl.create(:user, :with_mail, :admin => false)
+    FactoryGirl.create(:filter, :role => User.current.roles.first, :permissions => Permission.where(:name => 'view_hosts'), :search => "name = does_not_exist")
+    session = SETTINGS[:login] ? {:user => User.current.id, :expires_at => 5.minutes.from_now} : {}
+    get :host_template, {:kind => 'PXELinux', :spoof => @rh_host.ip, :format => 'text'}, session
+    assert_response :not_found
+    assert_template nil
```
Comment (domcleal) [inline test/functional/unattended_controller_test.rb]: ditto

File: test/functional/unattended_controller_test.rb [hunk 11]
```diff
@@ -358,6 +358,47 @@ class UnattendedControllerTest < ActionController::TestCase
...
+    session = SETTINGS[:login] ? {:user => User.current.id, :expires_at => 5.minutes.from_now} : {}
+    get :host_template, {:kind => 'PXELinux', :spoof => @rh_host.ip, :format => 'text'}, session
+    assert_response :success
+    assert @response.body.include?("linux")
+  end
+
+  test 'should not render a template to user with valid filter' do
+    User.current = FactoryGirl.create(:user, :with_mail, :admin => false)
+    FactoryGirl.create(:filter, :role => User.current.roles.first, :permissions => Permission.where(:name => 'view_hosts'), :search => "name = does_not_exist")
+    session = SETTINGS[:login] ? {:user => User.current.id, :expires_at => 5.minutes.from_now} : {}
```
Comment (domcleal) [inline test/functional/unattended_controller_test.rb]: And ditto in all of these tests, the same comments about using the existing `set_session_user`.

File: config/routes.rb [hunk 12]
```diff
@@ -419,12 +419,12 @@
...
   # get only for alterator unattended scripts
-  get 'unattended/provision/:metadata', :controller => 'unattended', :action => 'host_template', :format => 'html',
+  get 'unattended/provision/:metadata', :controller => 'unattended', :action => 'host_template', :format => 'text',
     :constraints => { :metadata => /(autoinstall\.scm|vm-profile\.scm|pkg-groups\.tar)/ }
   # get for end of build action
-  get 'unattended/built/(:id(:format))', :controller => 'unattended', :action => 'built'
+  get 'unattended/built/(:id(:format))', :controller => 'unattended', :action => 'built', :format => 'text'
   # get for all unattended scripts
-  get 'unattended/(:kind/(:id(:format)))', :controller => 'unattended', :action => 'host_template'
+  get 'unattended/(:kind/(:id(:format)))', :controller => 'unattended', :action => 'host_template', :format => 'text'
```
Comment (domcleal) [inline config/routes.rb:432]: I think `set_content_type` in the UnattendedController is now obsolete by setting format here, it should be removed as it appears to be overriding this setting.

File: app/controllers/application_controller.rb [hunk 13]
```diff
@@ -73,8 +73,17 @@ def set_gettext_locale_db
...
     if User.current && !User.current.hidden? && User.current.mail.blank?
-      notice _("Email is Required")
-      redirect_to edit_user_path(:id => User.current)
+      msg = _("Email is Required")
+      respond_to do |format|
+        format.html do
+          notice msg
+          redirect_to edit_user_path(:id => User.current)
+        end
+        format.any do
```
Comment (domcleal) [inline app/controllers/application_controller.rb]: format.text?

File: app/controllers/unattended_controller.rb (hunk 5 above)
Comment (lzap) [inline app/controllers/unattended_controller.rb]: Completely refactored how errors are reported, fixed above comment as well.

File: config/routes.rb (hunk 12 above)
Comment (lzap) [inline config/routes.rb:432]: Removed, there's an functional test for this so we should be safe.

Comment (lzap) [inline config/routes.rb:432]: Does not seem to work:

```
1) Failure:
UnattendedControllerTest#test_0008_unattended files content type should be text/plain [test/functional/unattended_controller_test.rb:89]:
Failed assertion, no message given.
```

Even tried `defaults: {:format => 'text'}` without any luck. I don't want to mess around, keeping the `set_content_type` there.

File: test/functional/unattended_controller_test.rb (hunk 8 above)
Comment (lzap) [inline test/functional/unattended_controller_test.rb]: I am not sure if I want to use fixtures, it's a step back actually. Time to provide `set_session_user` for FG objects?

File: test/functional/unattended_controller_test.rb (hunk 9 above)
Comment (lzap) [inline test/functional/unattended_controller_test.rb]: Sure.

**Comment by lzap** [comment]
Rebased, there is an open question for tests.

File: test/functional/unattended_controller_test.rb (hunk 8 above)
Comment (domcleal) [inline test/functional/unattended_controller_test.rb]: Yes, that's precisely what my comment shows - set_session_user can simply take a User.

**Comment by domcleal** [comment]
There are substantial numbers of test failures, the app is probably failing to startup.

File: test/functional/unattended_controller_test.rb (hunk 8 above)
Comment (lzap) [inline test/functional/unattended_controller_test.rb]: > Yes, that's precisely what my comment shows - set_session_user can simply take a User.

But it takes a fixture user, I am not going to add a new one. I need
special cases with/without email admin/non-admin.

I can perhaps modify the helper to accept User class instances, but
definitely not new fixture.

## 

Later,
 Lukas #lzap Zapletal

Comment (domcleal) [inline test/functional/unattended_controller_test.rb]: > I can perhaps modify the helper to accept User class instances, but

Yes, please do and just pass the user in - the code sample was in my first comment!

Comment (lzap) [inline test/functional/unattended_controller_test.rb]: > Yes, please do and just pass the user in - the code sample was in my first comment!

Sorry, I completely missed that comment :-( Will do.

## 

Later,
 Lukas #lzap Zapletal

**Comment by lzap** [comment]
Fixed all the four tests.

**Comment by domcleal** [comment]
All of the db:seed related tests are failing with:

<pre>
13:24:21 NameError: uninitialized constant TemplateKind::PREDEFINED
13:24:21 /var/lib/workspace/workspace/test_branch_upgrade/database/postgresql/ruby/2.0.0/db/seeds.d/07-provisioning_templates.rb:9:in `<top (required)>'
13:24:21 /var/lib/workspace/workspace/test_branch_upgrade/database/postgresql/ruby/2.0.0/db/seeds.rb:81:in `block in <top (required)>'
13:24:21 /var/lib/workspace/workspace/test_branch_upgrade/database/postgresql/ruby/2.0.0/db/seeds.rb:79:in `each'
13:24:21 /var/lib/workspace/workspace/test_branch_upgrade/database/postgresql/ruby/2.0.0/db/seeds.rb:79:in `<top (required)>'
13:24:21 /usr/local/rvm/gems/ruby-2.0.0-p353@test_branch_upgrade-3/gems/railties-4.2.6/lib/rails/engine.rb:547:in `load_seed'
13:24:21 /usr/local/rvm/gems/ruby-2.0.0-p353@test_branch_upgrade-3/gems/activerecord-4.2.6/lib/active_record/tasks/database_tasks.rb:250:in `load_seed'
13:24:21 /usr/local/rvm/gems/ruby-2.0.0-p353@test_branch_upgrade-3/gems/activerecord-4.2.6/lib/active_record/railties/databases.rake:183:in `block (2 levels) in <top (required)>'
13:24:21 /usr/local/rvm/gems/ruby-2.0.0-p353@test_branch_upgrade-3/bin/ruby_executable_hooks:15:in `eval'
13:24:21 /usr/local/rvm/gems/ruby-2.0.0-p353@test_branch_upgrade-3/bin/ruby_executable_hooks:15:in `<main>'
13:24:21 Tasks: TOP => db:seed
</pre>

**Comment by lzap** [comment]
Thanks for heads up, this was caused by merge conflict manual resolution.

File: app/controllers/unattended_controller.rb [hunk 14]
```diff
@@ -64,9 +70,8 @@ def render_template(type)
       logger.debug "rendering DB template #{config.name} - #{type}"
       safe_render config
     else
-      msg = "unable to find #{type} template for [#{@host.name}] running [#{@host.operatingsystem}]"
-      logger.error msg
-      render :text => msg, :status => :not_found
+      error_message = N_("unable to find %{type} template for %{host} running %{os}") % {:type => type, :host => @host.name, :os => @host.operatingsystem}
```
Comment (domcleal) [inline app/controllers/unattended_controller.rb]: Parameters shouldn't be interpolated before the string's passed to `_()` inside `render_custom_error` or the translation lookup will fail, they need to be done after translation.

Comment (lzap) [inline app/controllers/unattended_controller.rb]: Fixed, on second thought, unattended controller have i18n with defaults I guess (no session), but this will work with preview I guess.

File: app/controllers/unattended_controller.rb [hunk 15]
```diff
@@ -56,6 +56,12 @@ def require_ssl?
 
   private
 
+  def render_custom_error(status, error_message, params)
+    logger.error error_message
```
Comment (domcleal) [inline app/controllers/unattended_controller.rb]: error_message parameters must be interpolated before logging the message

**Comment by lzap** [comment]
Right, added and rebased on top of develop.

**Comment by domcleal** [comment]
On re-reviewing this, it seems the issue has already been fixed in b1997f52 when I refactored the controller in the same way that was suggested in the #10689 comments. Previewing templates as a non-admin user works fine.

It does though mean we're vulnerable in develop to the security issue I mentioned at https://github.com/theforeman/foreman/pull/2428#issuecomment-114842874, so this needs addressing.  I've filed this as http://projects.theforeman.org/issues/15490, and updated foreman-security + requested a CVE.

What I'd like to do is:
1. Close #10689 as a resolved or a dupe of #13039, which fixes the issue.
2. Update this PR to reference #15490, as it contains the authorisation fixes and fixes the error handling on 403s.
3. Update this PR to fix some review comments which I'll add shortly.

Sorry for the ticket change, but I'd like to be absolutely clear about what this is fixing.

File: app/services/foreman/access_permissions.rb [hunk 16]
```diff
@@ -351,7 +351,7 @@
                                                :externalNodes, :pxe_config, :storeconfig_klasses, :auto_complete_search, :bmc,
                                                :runtime, :resources, :templates, :overview, :nics],
                                     :dashboard => [:OutOfSync, :errors, :active],
-                                    :unattended => [:host_template, :hostgroup_template],
+                                    :unattended => [:host_template, :hostgroup_template] + TemplateKind.default_template_labels.keys.collect(&:to_sym),
```
Comment (domcleal) [inline app/services/foreman/access_permissions.rb]: This shouldn't be necessary now, since the action is always going to be either :host_template or :hostgroup_template - the routing changed in b1997f5 to route the template kinds to a single action.

File: test/functional/unattended_controller_test.rb [hunk 17]
```diff
@@ -358,12 +358,47 @@ class UnattendedControllerTest < ActionController::TestCase
...
+
+  test "should not render a template to user w/o email" do
+    user = FactoryGirl.create(:user)
+    get :host_template, {:kind => 'PXELinux', :spoof => @rh_host.ip, :format => 'text'}, set_session_user(user)
+    assert_response :unprocessable_entity
+  end
+
+  test "should not render a template to unauthorized user" do
+    user = FactoryGirl.create(:user, :with_mail)
+    user.auth_source.stubs(:authenticated?).returns(false)
```
Comment (domcleal) [inline test/functional/unattended_controller_test.rb]: I don't think this is needed or should be here, it looks like the test is meant to authenticate the user, but they shouldn't be authorised - not that authentication is failing.

File: test/functional/unattended_controller_test.rb [hunk 18]
```diff
@@ -358,12 +358,47 @@ class UnattendedControllerTest < ActionController::TestCase
...
+
+  test 'should render a template to user with valid filter' do
+    user = FactoryGirl.create(:user, :with_mail, :admin => false)
+    FactoryGirl.create(:filter, :role => user.roles.first, :permissions => Permission.where(:name => 'view_hosts'), :search => "name = #{@rh_host.name}")
+    get :host_template, {:kind => 'PXELinux', :spoof => @rh_host.ip, :format => 'text'}, set_session_user(user)
+    assert_response :success
+    assert @response.body.include?("linux")
+  end
+
+  test 'should not render a template to user with valid filter' do
```
Comment (domcleal) [inline test/functional/unattended_controller_test.rb]: _invalid_ filter

File: app/controllers/unattended_controller.rb [hunk 19]
```diff
@@ -56,6 +56,12 @@ def require_ssl?
 
   private
 
+  def render_custom_error(status, error_message, params)
+    logger.error error_message % params
+    # add a comment character (works with Red Hat and Debian systems) to avoid parsing errors
+    render(:text => '# ' + _(error_message) % params, :status => status)
```
Comment (domcleal) [inline app/controllers/unattended_controller.rb]: missing content-type? It's returned as text/html and formatted differently to the template itself.

**Comment by domcleal** [comment]
I'd like to add, that given the security issue, please let me know if you don't plan to update this PR to fix it and instead, I'll patch it myself - rather than leave it months again.

**Comment by domcleal** [comment]
@lzap if you could please update this soon to fix the security issue, or let me know if you don't plan to.

**Comment by lzap** [comment]
@domcleal had a huge github backlog and noticed just now. I can start tomorrow with this.

**Comment by lzap** [comment]
Amended all your last comments, rebased and pushed.

**Comment by domcleal** [comment]
Merged as c3c186d, thanks @lzap!
